src/server/transaction.cc

```diff
@@ -691,8 +691,12 @@ void Transaction::RunCallback(EngineShard* shard) {
   }
 
   // Log to journal only once the command finished running
-  if ((coordinator_state_ & COORD_CONCLUDING) || (multi_ && multi_->concluding))
+  if ((coordinator_state_ & COORD_CONCLUDING) || (multi_ && multi_->concluding)) {
     LogAutoJournalOnShard(shard, result);
+    if (cntx_) {
+      cntx_->conn_state.tracking_info_.Track(cntx_, cid_);
```

Comment:
Again, @dranikpg @kostasrim  would you mind providing an example or an explanation why triggering from the transaction is needed. I think this feature is eventually consistent by definition so I do not understand the reason for this change.

Reply:
Not sure what you mean... It has to be tracked with keys locked so we don't miss the next immediate update.

Technically it doesn't have to be tracked from transaction. We have the journal for that. But it has to be tracked during the transaction phase

Reply:
I do not argue, just trying to understand. What do you mean by "missing the next update" ?
can you please give me an example? suppose we won't do it withing the transaction phase, please provide the flow of steps that will lead to incorrect result as seen by a client.

Reply:
ok, so if A is being read and the rule is to track keys that are being read you should start tracking A immediately.

Got ya. But the whole code around this back and forth looks wrong:
Inside a shard callback we access connection state that is not designed to be accessed from shard threads.
Specifically, `Track` function checks  `if ((cid->opt_mask() & CO::READONLY) && cid->IsTransactional() && info.ShouldTrackKeys())` from the shard thread and then calls `AwaitFiberOnAll(std::move(cb));` on all shards.

Does it mean that if we call MGET on 20 shards, now all 20 shards call Track that checks if we should track and then calls back AwaitFiberOnAll on all shards, overall 20x20 calls?  Do I understand this interaction correctly?

Reply:
Ahh yes, after I asked Kostas to move the code, I didn't check it's correct 😅 We are already on the thread, we shouldn't be callin AwaitFiberOnAll

> Does it mean that if we call MGET on 20 shards, now all 20 sh

It should be the same as journaling. It is in fact journaling (or replication), just with a filter that sends invalidation except the new data